service/history/mutableStatePubSub.go

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/pborman/uuid"
+	"github.com/uber/cadence/common/persistence"
+)
+
+type (
+	// MutableStatePubSub is used to get updates on mutable state changes.
+	MutableStatePubSub interface {
+		Publish(*persistence.WorkflowMutableState)
+		Subscribe() (string, <-chan struct{})
```

Comment:
[not a blocker] consider returning a struct subscriber { id, ch }..  just makes the code read better on the caller. your unsubscribe can also take a subscriber as input

Reply:
Doing this is actually a little tricky because I want to return a read only channel to the caller. But if I return a subscriber struct then the channel has to be read&write so internally I can send on it.